hail/python/test/hail/backend/test_service_backend.py

```diff
@@ -9,20 +10,51 @@ def test_tiny_driver_has_tiny_memory():
     except Exception as exc:
         assert 'java.lang.OutOfMemoryError: Java heap space' in exc.args[0]
     else:
-        assert Fail
+        assert False
 
 @skip_unless_service_backend()
 def test_big_driver_has_big_memory():
-    old_driver_cores = hl.current_backend().driver_cores
-    old_driver_memory = hl.current_backend().driver_memory
+    backend = hl.current_backend()
+    assert isinstance(backend, ServiceBackend)
+    old_driver_cores = backend.driver_cores
+    old_driver_memory = backend.driver_memory
     try:
-        hl.current_backend().driver_cores = 8
-        hl.current_backend().driver_memory = 'highmem'
+        backend.driver_cores = 8
+        backend.driver_memory = 'highmem'
         t = hl.utils.range_table(100_000_000, 50)
         # The pytest (client-side) worker dies if we try to realize all 100M rows in memory.
         # Instead, we realize the 100M rows in memory on the driver and then take just the first 10M
         # rows back to the client.
         hl.eval(t.aggregate(hl.agg.collect(t.idx), _localize=False)[:10_000_000])
     finally:
-        hl.current_backend().driver_cores = old_driver_cores
-        hl.current_backend().driver_memory = old_driver_memory
+        backend.driver_cores = old_driver_cores
+        backend.driver_memory = old_driver_memory
+
+@skip_unless_service_backend()
+def test_tiny_worker_has_tiny_memory():
+    try:
+        t = hl.utils.range_table(1).annotate(nd=hl.nd.ones((30_000, 30_000)))
+        t = t.annotate(nd_sum=t.nd.sum())
+        t.aggregate(hl.agg.sum(t.nd_sum))
+    except Exception as exc:
+        assert 'HailException: Hail off-heap memory exceeded maximum threshold' in exc.args[0]
+    else:
+        assert False
+
+@skip_unless_service_backend()
+def test_big_worker_has_big_memory():
+    backend = hl.current_backend()
+    assert isinstance(backend, ServiceBackend)
+    old_driver_cores = backend.driver_cores
+    old_driver_memory = backend.driver_memory
+    try:
+        backend.worker_cores = 8
+        backend.worker_memory = 'highmem'
+        t = hl.utils.range_table(1).annotate(nd=hl.nd.ones((30_000, 30_000)))
+        t = t.annotate(nd_sum=t.nd.sum())
+        # We only eval the small thing so that we trigger an OOM on the worker
+        # but not the driver or client
+        t.aggregate(hl.agg.sum(t.nd_sum))
```

Comment:
@daniel-goldstein , Tim recently added an optimization that uses the driver machine if there's exactly one partition. In this instance, that violates your expectations about what a worker is.

This all feels a bit messy, but the time savings of running a single partition job on the driver seems worth the mess. If you switch this to `hl.utils.range_table(2, n_partitions=2)` the test will pass.

Reply:
Thanks! That saved me a world of confusion